File: frontend/src/auth/tokenManager.ts
import type { TokenResponse } from "../types";

const STORAGE_PREFIX = "oauth2_demo_";

let accessToken: string | null = null;
let refreshToken: string | null = null;

export function storeTokens(response: TokenResponse): void {
  accessToken = response.access_token;
  refreshToken = response.refresh_token ?? null;

  sessionStorage.setItem(`${STORAGE_PREFIX}access_token`, response.access_token);
  if (response.refresh_token) {
    sessionStorage.setItem(`${STORAGE_PREFIX}refresh_token`, response.refresh_token);
  }
  if (response.id_token) {
    sessionStorage.setItem(`${STORAGE_PREFIX}id_token`, response.id_token);
  }
}

export function getAccessToken(): string | null {
  return accessToken ?? sessionStorage.getItem(`${STORAGE_PREFIX}access_token`);
}

export function getRefreshToken(): string | null {
  return refreshToken ?? sessionStorage.getItem(`${STORAGE_PREFIX}refresh_token`);
}

export function hasStoredTokens(): boolean {
  return getAccessToken() !== null;
}

export function clearTokens(): void {
  accessToken = null;
  refreshToken = null;
  sessionStorage.removeItem(`${STORAGE_PREFIX}access_token`);
  sessionStorage.removeItem(`${STORAGE_PREFIX}refresh_token`);
  sessionStorage.removeItem(`${STORAGE_PREFIX}id_token`);
}

export function isTokenExpired(token: string): boolean {
  const parts = token.split(".");
  if (parts.length !== 3) return true;

  let payload = parts[1]!;
  // Pad base64 string
  const remainder = payload.length % 4;
  if (remainder > 0) {
    payload += "=".repeat(4 - remainder);
  }

  try {
    const json = JSON.parse(atob(payload)) as { exp?: number };
    if (typeof json.exp !== "number") return true;
    // Consider expired 30 seconds before actual expiry (same as iOS)
    return Date.now() / 1000 >= json.exp - 30;
  } catch {
    return true;
  }
}

File: ios/OAuth2Demo/Services/TokenManager.swift
import Foundation

actor TokenManager {
    private let keychain = KeychainService()
    private var refreshTask: Task<String, Error>?

    func storeTokens(_ response: TokenResponse) {
        keychain.save(key: .accessToken, value: response.accessToken)
        if let refreshToken = response.refreshToken {
            keychain.save(key: .refreshToken, value: refreshToken)
        }
        if let idToken = response.idToken {
            keychain.save(key: .idToken, value: idToken)
        }
    }

    func validAccessToken() async throws -> String {
        guard let accessToken = keychain.load(key: .accessToken) else {
            throw AuthError.tokenRefreshFailed
        }

        if !isTokenExpired(accessToken) {
            return accessToken
        }

        return try await refreshAccessToken()
    }

    func hasStoredTokens() -> Bool {
        keychain.load(key: .accessToken) != nil
    }

    func clearTokens() {
        keychain.deleteAll()
        refreshTask?.cancel()
        refreshTask = nil
    }

    private func refreshAccessToken() async throws -> String {
        if let existingTask = refreshTask {
            return try await existingTask.value
        }

        let task = Task<String, Error> {
            defer { refreshTask = nil }

            guard let refreshToken = keychain.load(key: .refreshToken) else {
                throw AuthError.tokenRefreshFailed
            }

            var request = URLRequest(url: AuthConfiguration.tokenEndpoint)
            request.httpMethod = "POST"
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

            let body = [
                "grant_type": "refresh_token",
                "client_id": AuthConfiguration.clientID,
                "refresh_token": refreshToken,
            ]
            request.httpBody = body
                .map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
                .joined(separator: "&")
                .data(using: .utf8)

            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                throw AuthError.tokenRefreshFailed
            }

            let tokenResponse = try JSONDecoder().decode(TokenResponse.self, from: data)
            storeTokens(tokenResponse)
            return tokenResponse.accessToken
        }

        refreshTask = task
        return try await task.value
    }

    private func isTokenExpired(_ token: String) -> Bool {
        let parts = token.split(separator: ".")
        guard parts.count == 3 else { return true }

        var payload = String(parts[1])
        // Pad base64 string
        let remainder = payload.count % 4
        if remainder > 0 {
            payload += String(repeating: "=", count: 4 - remainder)
        }

        guard let data = Data(base64Encoded: payload),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let exp = json["exp"] as? TimeInterval else {
            return true
        }

        // Consider expired 30 seconds before actual expiry
        return Date().timeIntervalSince1970 >= (exp - 30)
    }
}
